src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/StreamExtensions.cs

```diff
@@ -76,5 +76,35 @@ internal static int TryReadAll(this Stream stream, byte[] buffer, int offset, in
             }
             return totalBytesRead;
         }
+
+        /// <summary>
+        /// Resolve image size as either the given user-specified size or distance from current position to end-of-stream.
+        /// Also performs the relevant argument validation and publicly visible caller has same argument names.
+        /// </summary>
+        /// <exception cref="ArgumentException">size is 0 and distance from current position to end-of-stream can't fit in Int32.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">Size is negative or extends past the end-of-stream from current position.</exception>
+        internal static int GetAndValidateSize(Stream stream, int size, string streamParameterName)
+        {
+            long maxSize = stream.Length - stream.Position;
+
+            if (size != 0)
+            {
+                if (size > maxSize)
```

Comment:
Don't we need to handle size < 0 here? I think that's what previous unchecked was trying to do but left out cast accidentaly.